### PR TITLE
Add SASS support & Extracting stylesheets to seperate files

### PR DESCRIPTION
**Pull request by ottah**
- SASS/SCSS support
- Option to extracting stylesheets to named files (or keep the main bundle name). If not used the default behaviour will be the same as before and the styles will be bundled inside the js.
- Updated demo to test bundling

### CODE REVIEW

File: src/webpack/webpackTools.ts [hunk 1]
```diff
@@ -27,17 +28,38 @@ export interface CustomWebpackOptions extends webpack.Options.Optimization {
...
 }
 
 /**
  * Shared function to create a webpack config for an entry point
  */
+
 export function createConfig(outPath: string, entryPoint: string | webpack.Entry, options: CustomWebpackOptions, ...plugins: webpack.Plugin[]) {
+    let extractCSS = false;
+    if (options && options.extractStyles) {
+        extractCSS = options.extractStyles.extract || false;
```
Comment (rypete) [inline src/webpack/webpackTools.ts]: Are you able to squeeze this down to be
```
const extractCSS = options && options.extractStyles && options.extractStyles.extract
```

**Review by rypete — APPROVED** [review]

File: demo/src/client.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import './base.css';
+import './client.scss';
+
+console.log(1);
```
Comment (bryangaleopenfin) [inline demo/src/client.ts]: Can this be a more descriptive message?

**Comment by bryangaleopenfin** [comment]
What was the motivation for this? Is there a particular story this relates to?

**Review by bryangaleopenfin — COMMENTED** [review]
Struggling to get this to run

`npm i; npm run demo` works on develop but fails with an error on this branch -

    ERROR in ./src/base.css 1:5
    Module parse failed: Unexpected token (1:5)
    You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
    > html {
    |     background: black;
    |     color: red;
     @ ./src/client.ts 3:0-21

Any ideas? Maybe chat about this PR when you have a chance?

File: demo/res/demo/index.html [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
-</head>
-<body>
-    <p>demo</p>
-    <script src="main-bundle.js"></script>
-</body>
-</html>
\ No newline at end of file
+
+    <head>
+        <title>Test</title>
```
Comment (bryangaleopenfin) [inline demo/res/demo/index.html]: Suggest tidy this and `provider.html` if you're changing these:

So `<title>Test Provider</title>` and  `<title>Test Client</title>`, and similar unifying for the body

File: src/webpack/webpackTools.ts [hunk 4]
```diff
@@ -27,17 +28,27 @@ export interface CustomWebpackOptions extends webpack.Options.Optimization {
...
 /**
  * Shared function to create a webpack config for an entry point
  */
-export function createConfig(outPath: string, entryPoint: string | webpack.Entry, options: CustomWebpackOptions, ...plugins: webpack.Plugin[]) {
+
+export function createConfig(outPath: string, entryPoint: string | webpack.Entry, options?: CustomWebpackOptions, ...plugins: webpack.Plugin[]) {
+    let extractCSS = false;
+    if (options && options.extractStyles) {
+        extractCSS = !!options.extractStyles || false;
+    }
```
Comment (bryangaleopenfin) [inline src/webpack/webpackTools.ts]: Seconding Ryan's comment about making this one line so `extractCSS` can be `const`. Also not sure `|| false` will do anything in this case, given we've already got the `!!` cast to boolean

File: src/webpack/webpackTools.ts [hunk 5]
```diff
@@ -48,9 +59,38 @@ export function createConfig(outPath: string, entryPoint: string | webpack.Entry
         },
         module: {
             rules: [
-                {test: /\.css$/, loader: 'style-loader'},
-                {test: /\.css$/, loader: 'css-loader'},
-                {test: /\.module.css$/, loader: 'css-loader', query: {modules: true, localIdentName: '[name]__[local]___[hash:base64:5]'}},
+                {
+                    test: /\.module\.(sc|sa|c)ss$/,
+                    loader: [
+                        finalCSSLoader(extractCSS),
```
Comment (bryangaleopenfin) [inline src/webpack/webpackTools.ts]: Change `*CSS` to `*Css`? Think we're more often using `Url` etc. in code, and consistent with `MiniCssExtractPlugin`

File: src/webpack/webpackTools.ts [hunk 6]
```diff
@@ -66,10 +106,18 @@ export function createConfig(outPath: string, entryPoint: string | webpack.Entry
...
     }
+
     if (plugins && plugins.length) {
         config.plugins!.push.apply(config.plugins, plugins);
     }
 
+    if (extractCSS) {
+        const name = options!.extractStyles!;
+        config.plugins!.push(new MiniCssExtractPlugin({
+            filename: (typeof name === 'string') ? `${name}.css` : '[name].css'
```
Comment (bryangaleopenfin) [inline src/webpack/webpackTools.ts:114]: Is "[name]" something MiniCssExtractPlugin knows how to parse?

File: demo/src/index.ts [hunk 7]
```diff
@@ -1,4 +1,8 @@
+import './test.scss';
+import './base.css';
+
 declare const PACKAGE_VERSION: string;
 
 console.log('I am a test build file');
 console.log(PACKAGE_VERSION);
+
```
Comment (bryangaleopenfin) [inline demo/src/index.ts]: Where can I view this?

**Review by bryangaleopenfin — COMMENTED** [review]
Looks good, just a couple of questions and tidying suggestions

File: src/webpack/webpackTools.ts (hunk 6 above)
Comment (ottah) [reply src/webpack/webpackTools.ts:114]: Yeah the name gets passed to it from webpack just like the `output`: `[name]-bundle.js`

File: demo/src/index.ts (hunk 7 above)
Comment (ottah) [reply demo/src/index.ts]: `npm run demo` or `npm run demo:build`

File: src/webpack/webpackTools.ts (hunk 5 above)
Comment (bryangaleopenfin) [reply src/webpack/webpackTools.ts]: Same change for `extractCSS`

File: demo/src/index.ts (hunk 7 above)
Comment (bryangaleopenfin) [reply demo/src/index.ts]: I see the provider and client windows, but not this one?

Comment (ottah) [reply demo/src/index.ts]: Sorry had multiple branches testing different things with the demo folder. Should be fixed now.

File: src/webpack/webpackTools.ts (hunk 5 above)
Comment (ottah) [reply src/webpack/webpackTools.ts]: Renamed.

**Review by bryangaleopenfin — APPROVED** [review]

**Pull request merged**